File: packages/typescript/src/features/trust/providers/roles.ts
import { createUniqueUuid } from "../../../entities.ts";
import { logger } from "../../../logger.ts";
import {
	ChannelType,
	type IAgentRuntime,
	type Memory,
	type Provider,
	type ProviderResult,
	type State,
	type UUID,
} from "../../../types/index.ts";

export const roleProvider: Provider = {
	name: "ROLES",
	description:
		"Roles in the server, default are OWNER, ADMIN and MEMBER (as well as NONE)",
	dynamic: true,
	get: async (
		runtime: IAgentRuntime,
		message: Memory,
		state: State,
	): Promise<ProviderResult> => {
		const room = state.data.room ?? (await runtime.getRoom(message.roomId));
		if (!room) {
			logger.debug("No room found for roles provider, skipping");
			return {
				data: { roles: [] },
				values: {
					roles: "No room context available for role information.",
				},
				text: "No room context available for role information.",
			};
		}

		if (room.type !== ChannelType.GROUP) {
			return {
				data: {
					roles: [],
				},
				values: {
					roles:
						"No access to role information in DMs, the role provider is only available in group scenarios.",
				},
				text: "No access to role information in DMs, the role provider is only available in group scenarios.",
			};
		}

		const serverId = room.messageServerId;

		if (!serverId) {
			throw new Error("No server ID found");
		}

		logger.info(`Using server ID: ${serverId}`);

		const worldId = createUniqueUuid(runtime, serverId);
		const world = await runtime.getWorld(worldId);

		if (!world?.metadata?.ownership?.ownerId) {
			logger.info(
				`No ownership data found for server ${serverId}, initializing empty role hierarchy`,
			);
			return {
				data: {
					roles: [],
				},
				values: {
					roles: "No role information available for this server.",
				},
				text: "No role information available for this server.",
			};
		}

		const roles = world.metadata.roles || {};

		if (Object.keys(roles).length === 0) {
			logger.info(`No roles found for server ${serverId}`);
			return {
				data: {
					roles: [],
				},
				values: {
					roles: "No role information available for this server.",
				},
			};
		}

		logger.info(`Found ${Object.keys(roles).length} roles`);

		const owners: { name: string; username: string; names: string[] }[] = [];
		const admins: { name: string; username: string; names: string[] }[] = [];
		const members: { name: string; username: string; names: string[] }[] = [];

		for (const entityId of Object.keys(roles) as UUID[]) {
			const userRole = roles[entityId];

			const user = await runtime.getEntityById(entityId);

			const metadata = user?.metadata as
				| Record<string, Record<string, unknown>>
				| undefined;
			const name =
				(metadata?.default?.name as string) ||
				(metadata?.discord?.name as string) ||
				user?.names?.[0];

			const username =
				(metadata?.default?.username as string) ||
				(metadata?.discord?.username as string) ||
				(metadata?.discord?.userName as string) ||
				user?.names?.[0];

			const names = user?.names as string[];

			if (
				owners.some((owner) => owner.username === username) ||
				admins.some((admin) => admin.username === username) ||
				members.some((member) => member.username === username)
			) {
				continue;
			}

			if (!name || !username || !names || names.length === 0) {
				logger.warn(
					`User ${entityId} (role: ${userRole}) has no name or username, skipping`,
				);
				continue;
			}

			switch (userRole) {
				case "OWNER":
					owners.push({ name, username, names });
					break;
				case "ADMIN":
					admins.push({ name, username, names });
					break;
				default:
					members.push({ name, username, names });
					break;
			}
		}

		let response = "# Server Role Hierarchy\n\n";

		if (owners.length > 0) {
			response += "## Owners\n";
			owners.forEach((owner) => {
				response += `${owner.name} (${owner.names.join(", ")})\n`;
			});
			response += "\n";
		}

		if (admins.length > 0) {
			response += "## Administrators\n";
			admins.forEach((admin) => {
				response += `${admin.name} (${admin.names.join(", ")}) (${admin.username})\n`;
			});
			response += "\n";
		}

		if (members.length > 0) {
			response += "## Members\n";
			members.forEach((member) => {
				response += `${member.name} (${member.names.join(", ")}) (${member.username})\n`;
			});
		}

		if (owners.length === 0 && admins.length === 0 && members.length === 0) {
			return {
				data: {
					roles: [],
				},
				values: {
					roles: "No role information available for this server.",
				},
				text: "No role information available for this server.",
			};
		}

		return {
			data: {
				roles: response,
			},
			values: {
				roles: response,
			},
			text: response,
		};
	},
};

export default roleProvider;

File: packages/python/elizaos/features/core_capabilities/trust/providers/roles.py
"""Roles provider.

Provides the server role hierarchy (Owner, Admin, Member) for the
current room/world context. Ported from the TypeScript ``roleProvider``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


async def get_roles(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    """Provide the server role hierarchy for the current room."""
    # Retrieve room
    get_room = getattr(runtime, "get_room", None)
    room = await get_room(message.room_id) if callable(get_room) else None

    if room is None:
        return ProviderResult(
            text="No room context available for role information.",
            values={"roles": "No room context available for role information."},
            data={"roles": []},
        )

    # Only available in group contexts
    room_type = getattr(room, "type", None) or getattr(room, "channel_type", None)
    if room_type and str(room_type).upper() not in ("GROUP", "WORLD"):
        return ProviderResult(
            text="No access to role information in DMs, the role provider is only available in group scenarios.",
            values={
                "roles": "No access to role information in DMs, the role provider is only available in group scenarios.",
            },
            data={"roles": []},
        )

    server_id = getattr(room, "message_server_id", None) or getattr(room, "messageServerId", None)
    if not server_id:
        return ProviderResult(
            text="No role information available for this server.",
            values={"roles": "No role information available for this server."},
            data={"roles": []},
        )

    # Retrieve world
    world_id = getattr(room, "world_id", None) or getattr(room, "worldId", None)
    if not world_id:
        # Try to construct from server_id
        create_uuid = getattr(runtime, "create_unique_uuid", None)
        world_id = create_uuid(server_id) if callable(create_uuid) else server_id

    get_world = getattr(runtime, "get_world", None)
    world = await get_world(world_id) if callable(get_world) else None

    if world is None:
        return ProviderResult(
            text="No role information available for this server.",
            values={"roles": "No role information available for this server."},
            data={"roles": []},
        )

    metadata: dict[str, Any] = getattr(world, "metadata", None) or {}
    ownership = metadata.get("ownership", {})
    if not ownership.get("ownerId"):
        return ProviderResult(
            text="No role information available for this server.",
            values={"roles": "No role information available for this server."},
            data={"roles": []},
        )

    roles: dict[str, str] = metadata.get("roles", {})
    if not roles:
        return ProviderResult(
            text="No role information available for this server.",
            values={"roles": "No role information available for this server."},
            data={"roles": []},
        )

    # Categorize entities by role
    owners: list[dict[str, str]] = []
    admins: list[dict[str, str]] = []
    members: list[dict[str, str]] = []

    get_entity = getattr(runtime, "get_entity_by_id", None)

    for entity_id, user_role in roles.items():
        # Try to resolve entity name
        name = entity_id
        if callable(get_entity):
            try:
                entity = await get_entity(entity_id)
                if entity:
                    names = getattr(entity, "names", None) or []
                    entity_meta = getattr(entity, "metadata", None) or {}
                    default_meta = (
                        entity_meta.get("default", {}) if isinstance(entity_meta, dict) else {}
                    )
                    name = default_meta.get("name") or (names[0] if names else None) or entity_id
            except Exception:
                pass

        entry = {"name": str(name), "entityId": entity_id, "role": user_role}
        role_upper = user_role.upper() if user_role else ""
        if role_upper == "OWNER":
            owners.append(entry)
        elif role_upper == "ADMIN":
            admins.append(entry)
        else:
            members.append(entry)

    # Build response text
    parts: list[str] = ["# Server Role Hierarchy\n"]

    if owners:
        parts.append("## Owners")
        for o in owners:
            parts.append(f"{o['name']} ({o['entityId']})")
        parts.append("")

    if admins:
        parts.append("## Administrators")
        for a in admins:
            parts.append(f"{a['name']} ({a['entityId']})")
        parts.append("")

    if members:
        parts.append("## Members")
        for m in members:
            parts.append(f"{m['name']} ({m['entityId']})")

    if not owners and not admins and not members:
        return ProviderResult(
            text="No role information available for this server.",
            values={"roles": "No role information available for this server."},
            data={"roles": []},
        )

    response_text = "\n".join(parts)

    return ProviderResult(
        text=response_text,
        values={"roles": response_text},
        data={"roles": response_text},
    )


role_provider = Provider(
    name="ROLES",
    description="Roles in the server, default are OWNER, ADMIN and MEMBER (as well as NONE)",
    get=get_roles,
    dynamic=True,
)
